lib/Dialect/Triton/IR/Traits.cpp

```diff
@@ -142,7 +143,12 @@ LogicalResult OpTrait::impl::verifyTensorLayouts(Operation *op) {
                          << moduleThreadsPerWarp << " threads per warp.";
       }
 
-      int moduleWarpsPerCTA = ttg::TritonGPUDialect::getNumWarps(module);
+      int moduleWarpsPerCTA =
+        // TODO: Determine why getNumWarps is called in a LLVM_DEBUG block
+        //       before the attribute triton_gpu.num-warps is set.
+        (triton::tools::getBoolEnv("TRITON_ENABLE_LLVM_DEBUG") &&
+         !module->hasAttr("triton_gpu.num-warps")) ? 0 :
+      ttg::TritonGPUDialect::getNumWarps(module);
```

Comment:
I don't understand why this change is needed. Can we figure this out before merging this PR? otherwise I feel like the TODO will never get done.

Reply:
That's fair.

I need to dig a little deeper on this one, but what I know so far is that getNumWarps hits a llvm::report_fatal_error when the triton_gpu.num-warps attribute is not set.

I think somewhere in Triton there is probably some debug output that is hitting verifyTensorLayouts before the num-warps attribute is set, and that it is just not being hit because the code path isn't covered.

Reply:
yup; we probably have a small bug somewhere and this seems like a good time to hunt it :)

Reply:
So far it looks like the root of the difference in behavior is in llvm-project/mlir itself at logSuccessFullPatternApplication:

https://github.com/llvm/llvm-project/blob/10c48a772742b7afe665a815b7eba2047f17dc4b/mlir/lib/Rewrite/PatternApplicator.cpp#L223

This function is called inside a LLVM_DEBUG block. The reason for the llvm::report_fatal_error in DebugFlag=true mode is that Triton doesn't set num_warps is set at https://github.com/openai/triton/blob/de5528c3e74b74e7aa528fc3715b071a6e095b91/lib/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.cpp#L802 in ConvertTritonToTritonGPU.

So anywhere the verifier code from the logSuccessFullPatternApplication invocation that happens before ConvertTritonToTritonGPU seems to hit this case. 


If I set DebugFlag to true after the ConvertTritonToTritonGPU pass is run, I don't observe the fatal error. Do you think it makes sense to change how getNumWarps or verifyTensorLayouts  behave before the ConvertTritonToTritonGPU type passes are run?

Reply:
I have added a change that might fix a potential Triton bug into this PR. I noticed that the fatal error happens when calling applyPartialConversion from ConvertTritonToTritonGPU. The callsite is actually just before attributes like num-warps, wraps-per-thread and num-ctas are set. I changed it to call the applyPartialConversion just after those are set. 

Does that seem like the right approach or bug fix to you @ptillet ?

Reply:
yep, that works